pkg/server/plugin/upstreamauthority/certmanager/certmanager.go

```diff
@@ -0,0 +1,264 @@
+	}
+	// If no issuer_kind given, default to Issuer
+	if len(config.IssuerKind) == 0 {
+		p.log.Warn("configuration has empty issuer_kind property, defaulting to 'Issuer'")
+		config.IssuerKind = "Issuer"
+	}
+	// If no issuer_group given, default to cert-manager.io
+	if len(config.IssuerGroup) == 0 {
+		p.log.Warn("configuration has empty issuer_group property, defaulting to 'cert-manager.io'")
+		config.IssuerGroup = "cert-manager.io"
+	}
+	if len(config.KubeConfigFilePath) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "configuration has empty kube_config_path property")
+	}
+
+	return config, nil
+}
+
+func (p *Plugin) MintX509CA(request *upstreamauthorityv0.MintX509CARequest, stream upstreamauthorityv0.UpstreamAuthority_MintX509CAServer) error {
+	ctx := stream.Context()
+
+	// Build PEM encoded CSR
+	csrBuf := new(bytes.Buffer)
+	err := pem.Encode(csrBuf, &pem.Block{
+		Type:  "CERTIFICATE REQUEST",
+		Bytes: request.Csr,
+	})
+	if err != nil {
+		return err
+	}
+
+	// Build CertificateRequest object to be created. Use a generated name to
+	// avoid conflicts.
+	cr := &cmapi.CertificateRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "spiffe-ca-",
+			Namespace:    p.config.Namespace,
+		},
+		Spec: cmapi.CertificateRequestSpec{
+			Duration: &metav1.Duration{
+				Duration: time.Second * time.Duration(request.PreferredTtl),
+			},
+			Request: csrBuf.Bytes(),
+			Usages: []cmapi.KeyUsage{
+				cmapi.UsageDigitalSignature,
+				cmapi.UsageKeyEncipherment,
+				cmapi.UsageCertSign,
+			},
+			IsCA: true,
+			IssuerRef: cmmeta.ObjectReference{
+				Name:  p.config.IssuerName,
+				Kind:  p.config.IssuerKind,
+				Group: p.config.IssuerGroup,
+			},
+		},
+	}
+	if err := p.cmclient.Create(ctx, cr); err != nil {
+		return err
+	}
+
+	// Poll the CertificateRequest until it is signed. If not signed after 300
+	// polls, error.
+	obj := client.ObjectKey{Name: cr.Name, Namespace: cr.Namespace}
+	for i := 0; true; i++ {
+		if i == 60*5 { // ~5 mins
+			return fmt.Errorf("failed to wait for CertificateRequest %s/%s to become ready: %#+v", cr.Namespace, cr.Name, cr.Status)
+		}
+
+		time.Sleep(time.Second)
+
+		if err := p.cmclient.Get(ctx, obj, cr); err != nil {
+			return err
+		}
+
+		if cmutil.CertificateRequestIsDenied(cr) {
+			cond := cmutil.GetCertificateRequestCondition(cr, cmapi.CertificateRequestConditionDenied)
+			return fmt.Errorf("created CertificateRequest %s/%s has been denied: %s:%s", cr.Namespace, cr.Name, cond.Reason, cond.Message)
+		}
+
+		if len(cr.Status.Certificate) > 0 {
+			break
+		}
+	}
+
+	// Parse signed certificate chain and CA certificate from CertificateRequest
+	caChain, err := pemutil.ParseCertificates(cr.Status.Certificate)
+	if err != nil {
+		return fmt.Errorf("failed to parse signed certificate from %s/%s: %w", obj.Namespace, obj.Name, err)
+	}
+
+	// If the configured issuer did not populate the CA on the request we cannot
+	// build the upstream roots. We can only error here.
```

Comment:
💯 

Is this something that can happen "normally"? Do we need to document anything about it e.g. our requirements on a certain cert-manager configuration?

Reply:
It is up to the cert-manager Issuer implementation as to whether they support setting the CA or not, and we can't programmatically determine this before as they are just stringy references.

For cert-manager core Issuers,Vault and CA support setting this field, but ACME for example doesn't.